### PR TITLE
Solmate proofs

### PR DESCRIPTION
**Pull request by doubledup**
Needed to add a new function to get the root hash, as the only one available is for verifying a proof. It's currently in our [solmate fork](https://github.com/Snowfork/solmate). I'll work on pushing this upstream.

Also need to check the licenses: Solmate is released under AGPL 3 (though the files we're using have an MIT license identifier as their first line) and our contracts under Apache 2.

Added a large proof test (with data generated in a [benchmarks repo](https://github.com/doubledup/merkle-proof-contract-benchmarks/blob/a14cdb5162eb5cfadfe4042961e716a249969901/src/bin/generate_proof.rs)) and a fuzz test.

### CODE REVIEW

**Review by vgeddes — APPROVED** [review]
Looks good! Lets merge it.

File: core/packages/contracts/src/InboundQueue.sol [hunk 1]
```diff
@@ -67,7 +67,7 @@ contract InboundQueue is AccessControl {
 
     function submit(Message calldata message, bytes32[] calldata leafProof, bytes calldata headerProof) external {
         bytes32 leafHash = keccak256(abi.encode(message));
```
Comment (vgeddes) [inline core/packages/contracts/src/InboundQueue.sol:69]: Lets add comment explaining the purpose of this hashing the leaf

File: core/packages/contracts/src/InboundQueue.sol [hunk 2]
```diff
@@ -67,7 +67,7 @@ contract InboundQueue is AccessControl {
...
-        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
+        bytes32 commitment = MerkleProofLib.calculateRoot(leafProof, leafHash);
```
Comment (vgeddes) [inline core/packages/contracts/src/InboundQueue.sol:70]: Ditto, lets add a comment explaining why we're calculating the root.

File: core/packages/contracts/src/InboundQueue.sol [hunk 3]
```diff
@@ -67,7 +67,7 @@ contract InboundQueue is AccessControl {
...
         if (!parachainClient.verifyCommitment(commitment, headerProof)) {
```
Comment (vgeddes) [inline core/packages/contracts/src/InboundQueue.sol:71]: Ditto, and another comment explaining what is actually being verified here.

**Review by vgeddes — APPROVED** [review]

**Review by alistair-singh — APPROVED** [review]
🚀

**Comment by yrong** [comment]
Are we missing the refactoring here?
https://github.com/Snowfork/snowbridge/blob/d001d7255ca1d8e81305795be19484e4a9fa42d3/core/packages/contracts/src/ParachainClient.sol#L4

**Comment by yrong** [comment]
Since we've introduce fuzz test here, would it better to explicit declare in foundry.toml like this?
https://github.com/transmissions11/solmate/blob/2001af43aedb46fdc2335d2a7714fb2dae7cfcd1/foundry.toml#L6-L7

**Review by yrong — APPROVED** [review]
Awesome!

**Comment by vgeddes** [comment]
> Are we missing refactoring here?
> 
> https://github.com/Snowfork/snowbridge/blob/d001d7255ca1d8e81305795be19484e4a9fa42d3/core/packages/contracts/src/ParachainClient.sol#L4

Yeah, good catch! @doubledup you need to update both ParachainClient and BeefyClient to use the new merkle proof verifier.

**Comment by doubledup** [comment]
The default number of fuzz test runs is 256, I've set it to 1000 in our default profile.

**Pull request merged**